modules/transport-netty4/src/main/java/org/opensearch/http/netty4/Netty4HttpServerTransport.java

```diff
@@ -413,10 +413,9 @@ protected void channelRead0(ChannelHandlerContext ctx, HttpMessage msg) throws E
                     // If this handler is hit then no upgrade has been attempted and the client is just talking HTTP
                     final ChannelPipeline pipeline = ctx.pipeline();
                     pipeline.addAfter(ctx.name(), "handler", getRequestHandler());
-                    pipeline.replace(this, "aggregator", aggregator);
+                    pipeline.replace(this, "decoder_compress", new HttpContentDecompressor());
```

Comment:
I haven't looked through this deeply yet so apologies in advance if this comment is not helpful...but I find this all this "replace", "addAfter", "addBefore", etc makes it very hard to reason about the state of the pipeline when this is done. I'm wondering if it is possible to simplify this.

One thing that jumps out is that we're adding handlers before "handler" (which is added above). Could that be re-written as the following?

```
final ChannelPipeline pipeline = ctx.pipeline();
pipeline.addAfter(ctx.name(), "request_creator", ...
pipeline.addAfter("request_creator", "response_creator", ...
pipeline.addAfter("response_creator", "pipelining", ...
pipeline.addAfter("pipelining", "handler", ...
```

Also sometimes we're using the `pipeline` local variable and sometimes we're using `ch.pipeline()`. I'm assuming those refer to the same instance?

In any case, making the Netty pipeline setup as simple as possible can really help with long term maintainability.

Reply:
Even better, if we know that `ctx.name()` is the last handler in the pipeline (this handler was just added via an addLast call so I think it should be), then you can just add the new handlers via `addLast` calls, which I find is the simplest way to construct a pipeline.

Reply:
Thanks @andrross 

> Also sometimes we're using the pipeline local variable and sometimes we're using ch.pipeline(). I'm assuming those refer to the same instance?

Yes, thanks for noticing, I will change that to use `pipeline` in this block

> One thing that jumps out is that we're adding handlers before "handler" (which is added above). Could that be re-written as the following where it's more of a simple chaining of handlers?

The order of handler is super important, and in this particular block - we really need to replace the handler and restart the processing again. The whole kind of purpose of that is to process the upgrade request first and than run the original request as it have been processed the normal way.

> Even better, if we know that ctx.name() is the last handler in the pipeline (this handler was just added via an addLast call so I think it should be), then you can just add the new handlers via addLast calls, which I find is the simplest way to construct a pipeline.

The ` ctx.name()` is the current handler which we are replacing with `decoder_compress`, there are a few other handlers added before and/or after (by means of HttpServerUpgradeHandler  and CleartextHttp2ServerUpgradeHandler).

Reply:
> The order of handler is super important

Understood. I'm trying to see if there's a more straightforward way to achieve the desired order so that the actual order is more apparent/easier to understand when reading the code. It looks to me light we add "handler" to the pipeline, then 10 lines later start adding handlers before it. Seems like that could all happen together, and potentially invert the order of adding them so that you don't have to mix "addAfter" and "addBefore".

Reply:
I think in general it is possible to rewrite the pipeline configuration a bit, it took me a while to understand why some handlers are either not invoked or invoked when they shouldn't, luckily we have test for it now :-) (not like affraid to touch it but it took me really a lot of time today to troubleshoot) :-)

Reply:
Fair enough...I did start this comment with "apologies in advance" :)

Reply:
> Fair enough...I did start this comment with "apologies in advance" :)

Thanks a lot for review